### PR TITLE
Various refactors in preparation for adding error reporting

### PR DESCRIPTION
**Pull request by nex3**
Sending these out on their own to avoid dumping another giant pull
request.

See #912

### CODE REVIEW

File: lib/src/solver/incompatibility_cause.dart [hunk 1]
```diff
@@ -0,0 +1,47 @@
+// Copyright (c) 2017, the Dart project authors.  Please see the AUTHORS file
```
Comment (natebosch) [inline lib/src/solver/incompatibility_cause.dart]: year += 1

File: lib/src/solver/failure.dart [hunk 2]
```diff
@@ -2,201 +2,22 @@
...
-  final Version version;
-  final String feature;
-
-  String get _message =>
-      "$package $version doesn't have a feature named $feature";
 
-  MissingFeatureException(String package, this.version, this.feature,
-      Iterable<Dependency> dependencies)
-      : super(package, dependencies);
+  String toString() => "Tough luck, Chuck!";
```
Comment (munificent) [inline lib/src/solver/failure.dart]: Add a TODO to put something real here?

File: lib/src/solver/incompatibility.dart [hunk 3]
```diff
@@ -44,39 +52,84 @@ class Incompatibility {
...
+      assert(terms.length == 1);
+      assert(terms.first.isPositive);
+
+      if (terms.first.constraint.isAny) {
+        return "no versions of ${_terseRef(terms.first)} "
+            "are compatible with the current SDK";
+      } else {
+        return "${terms.first.package.toTerseString()} is incompatible with "
+            "the current SDK";
+      }
```
Comment (munificent) [inline lib/src/solver/incompatibility.dart]: Might be useful to show the SDK version number here, no?

File: lib/src/solver/incompatibility.dart [hunk 4]
```diff
@@ -44,39 +52,84 @@ class Incompatibility {
...
+          return byRef.values;
+        }).toList(),
+        cause);
   }
 
-  Incompatibility._(this.terms);
+  Incompatibility._(this.terms, this.cause);
 
   String toString() {
+    if (cause == IncompatibilityCause.dependency) {
```
Comment (munificent) [inline lib/src/solver/incompatibility.dart]: What do you think of moving most of this logic into IncompatibilityCause? Something like:

```dart
class Incompability {
  String toString() => cause.describe(terms);
}

abstract class IncompatibilityCause {
  String describe(List<Term> terms);
}

class SdkCause implements IncompatibilityCause {
  String describe(List<Term> terms) {
    // ...
  }
}
```

?

File: lib/src/solver/failure.dart (hunk 2 above)
Comment (nex3) [reply lib/src/solver/failure.dart]: Done.

File: lib/src/solver/incompatibility.dart (hunk 4 above)
Comment (nex3) [reply lib/src/solver/incompatibility.dart]: It always weirds me out a bit to have a method like `Incompatibility.describe()` that's only ever supposed to be invoked by one specific caller... it feels like kind of the wrong separation of concerns.

File: lib/src/solver/incompatibility.dart (hunk 3 above)
Comment (nex3) [reply lib/src/solver/incompatibility.dart]: Added a TODO.

File: lib/src/solver/incompatibility_cause.dart (hunk 1 above)
Comment (nex3) [reply lib/src/solver/incompatibility_cause.dart]: Done.

**Pull request merged**